Added ability to override sender E-Mail

dmorlitz opened the pull request:
For some projects you may want to be able to accept incoming E-Mail from anyone.  I have updated the Mailgun plugin to allow you to specify on a per-project basis if you wish to use a specific E-Mail address for messages received from unknown senders.

If you leave this setting blank in the project integrations configuration, unknown senders are still ignored as they are now.

This enables mailing lists where the sender is not always known to post tasks if they forward messages to the correct E-Mail address for a project.

The user you specify for the override must have rights to the project still.  If the sender is already authorized to the project, their information is used before the override is put in......meaning the override is only used for unknown senders to Kanboard.

Commit:
Added ability for projects to override sender to catchall address

Commit:
Added ability to accept incoming mail from anyone on a per project basis

Commit:
Improved logging of mapped user

fguillot commented on Controller/MailgunProjectSettings.php:
```diff
@@ -0,0 +1,47 @@
+
+}
+
+?>
```
There is no need to have closing tag here.

dmorlitz replied:
Fixed

fguillot commented on README.md:
```diff
@@ -111,6 +111,9 @@ Notes
 -------
 
+### Version 1.0.10
+- For unknown senders you can now accept them into a project by specifying a user to accept the task - this is configurable on a project level - leaving the setting blank will cause unknown senders to be ignored as they are now
```
There is a missing blank line here.

dmorlitz replied:
Fixed

fguillot commented on Template/config/mailgun_id_mapping.php:
```diff
@@ -0,0 +1,12 @@
+<h3><img src="<?= $this->url->dir() ?>plugins/Mailgun/mailgun-icon.png"/>&nbsp;Mailgun Unknown ID mapping</h3>
```
This text should be translatable.

dmorlitz replied:
I have added the translation tag......do I need to provide the translations?

fguillot commented on EmailHandler.php:
```diff
@@ -95,22 +95,30 @@ public function validate(array $payload)
+        $user = $this->userModel->getByEmail($payload['sender']);
+
+	// Check to see if a catchall user was specified - if the original sender is unrecognized
+	if (empty($user)) {
```
There is some indentation issue here. Make sure it's consistent with the rest of the code.

dmorlitz replied:
fixed

fguillot commented on EmailHandler.php:
```diff
@@ -95,22 +95,30 @@ public function validate(array $payload)
-        }
-
         // The project must have a short name
+        //    DMM: Moved project check in front of user check - to support project unknown user mappings
```
Not sure if this comment is really useful.

dmorlitz replied:
Sorry - this was just my note taking - I have removed it

fguillot commented on Controller/MailgunProjectSettings.php:
```diff
@@ -0,0 +1,47 @@
+
+    public function save()
+    {
+
```
Useless blank line. Make sure to follow existing code style.

dmorlitz replied:
Fixed - I didn't even realize that I added anything to that function

fguillot reviewed (changes requested):
Please, update the unit tests according to your changes and fix the various coding style comments.

Commit:
Added unit tests and fixed some style issues

Commit:
Added unit tests - fixed PHP 5.6 project numbering in test - fixed some style issues

Commit:
Added unit tests - fixed PHP 5.6 project numbering in test - fixed some style issues

Commit:
use KanboardModelProjectMetadataModel;

Commit:
Added unit tests - fixed PHP 5.6 project numbering in test - fixed some style issues

Commit:
Added unit tests - fixed PHP 5.6 project numbering in test - fixed some style issues

Commit:
Added unit tests - fixed PHP 5.6 project numbering in test - fixed some style issues

Commit:
Added unit tests - fixed PHP 5.6 project numbering in test - fixed some style issues

Commit:
Added unit tests - fixed PHP 5.6 project numbering in test - fixed some style issues

dmorlitz commented:
I believe I have made all requested changes included adding the necessary unit test.  Thank you for your patience as I learn how this system works.

dmorlitz commented:
Were there any other changes you wanted in this request?  I think I have addressed all of the concerns you had - I just am not sure how to re-submit the pull request for review

fguillot commented on Template/config/mailgun_id_mapping.php:
```diff
@@ -0,0 +1,12 @@
+<div class="panel">
+    <?= $this->form->label(t('Set unknown senders to:'), 'MailgunProject_catchall') ?>
+    <?= $this->form->text('MailgunProject_catchall', $values) ?>
+    <?= $this->form->label(t('(NOTE: Blank value will cause unknown senders to be ignored)'), 'MailgunProject_catchall') ?>
```
This text should not be a form label. Instead, move this to the `form-help` paragraph below, before the link.

dmorlitz replied:
Text moved

fguillot commented on Plugin.php:
```diff
@@ -17,6 +17,7 @@ class Plugin extends Base
     public function initialize()
     {
         $this->emailClient->setTransport('mailgun', '\Kanboard\Plugin\Mailgun\EmailHandler');
+        $this->template->hook->attach('template:project:integrations', 'mailgun:config/mailgun_id_mapping');
```
This is probably better to put this template into a `project` folder (for example `project/integration`) instead of `config`. This is to stay consistent with the rest of the application `config` is for the global application settings.

fguillot commented on Controller/MailgunProjectSettings.php:
```diff
@@ -0,0 +1,43 @@
+        $this->response->html($this->helper->layout->project('MailgunProject:settings', array(
+            'owners' => $this->projectUserRoleModel->getAssignableUsersList($project['id'], true),
+	    'values' => array(
+                'MailgunProject_catchall'   => $this->projectMetadataModel->get($project['id'], 'MailgunProject_catchall'),
```
Please, don't mix camel case and snake case for the metadata key. Use something like `mailgun_catch_all` instead of `MailgunProject_catchall`. (`Project` is redundant since it's a project metadata).

fguillot commented on Template/config/mailgun_id_mapping.php:
```diff
@@ -0,0 +1,12 @@
+<h3><img src="<?= $this->url->dir() ?>plugins/Mailgun/mailgun-icon.png"/>&nbsp;<?= t('Mailgun Unknown ID mapping') ?></h3>
+<div class="panel">
+    <?= $this->form->label(t('Set unknown senders to:'), 'MailgunProject_catchall') ?>
```
You don't need to add the `:` at the end. This is to stay consistent with the rest.

dmorlitz replied:
Removed

fguillot commented on Locale/fr_FR/translations.php:
```diff
@@ -1,6 +1,7 @@
 <?php
 
 return array(
+    'Mailgun Unknown ID mapping' => 'Mappage d\'ID inconnu de Mailgun',
```
I will update the French translation, once it get merged. Don't worry about that.

dmorlitz replied:
I removed this file so you can add the translation

fguillot reviewed (changes requested):
This is mostly cosmetic changes. Thank you for taking them into consideration.

Commit:
Cosmetic changes

Commit:
Cosmetic changes

Commit:
Removed unnecessary duplicate tests

fguillot commented on Controller/MailgunProjectSettings.php:
```diff
@@ -0,0 +1,43 @@
+<?php
```
I just noticed that this file is not used at all. It can be removed.

fguillot commented on Locale/fr_FR/translations.php:
```diff
@@ -1,6 +0,0 @@
-<?php
```
Please, do not remove that file.

fguillot commented on Template/project/integration/mailgun_id_mapping.php:
```diff
@@ -0,0 +1,14 @@
+    <?= $this->form->text('mailgun_catch_all', $values) ?>
+
+    <p class="form-help">
+        <?= t('(NOTE: Blank value will cause unknown senders to be ignored)') ?><p>
```
This `<p>` tag should be removed.

fguillot commented on Test/EmailHandlerTest.php:
```diff
@@ -88,6 +89,7 @@ public function testHandlePayload()
         $taskFinderModel = new TaskFinderModel($this->container);
 
         $this->assertEquals(2, $userModel->create(array('username' => 'me', 'email' => 'me@localhost')));
+        $this->assertEquals(3, $userModel->create(array('username' => 'anyone', 'email' => 'anyone@localhost')));
```
This test case should not be altered.

fguillot commented on Test/EmailHandlerTest.php:
```diff
@@ -116,6 +118,40 @@ public function testHandlePayload()
+        $this->assertTrue($projectUserRoleModel->addUser(3, 3, Role::PROJECT_MEMBER));
+
+        // The task must be created
+        $this->assertTrue($emailHandler->receiveEmail(array('sender' => 'd@e.f', 'subject' => 'Email task', 'recipient' => 'test4@localhost', 'stripped-html' => '<strong>boo</strong>')));
```
I do not understand your test case, where `anyone@localhost` is used if not for the sender nor the recipient?

fguillot reviewed (commented):
Sorry, I clicked on the wrong button and I closed your PR. You might need to open a new PR. I noticed new things that need to be fixed.
